[wip,dnm] tracing: peephole optimizations

irfansharif opened the pull request.

irfansharif committed:
sql: revive BenchmarkTracing

Fixes #62081. It had rotted after we removed the trace.mode setting. It
now makes use of a debug-only cluster setting that if enabled, installs
real spans all sql statements. We'll defer the actual investigation into
the cause of the slow-down in future PRs.

    $ make bench PKG=./pkg/bench  TESTFLAGS='-benchtime=10000x -count 20' \
        BENCHES='BenchmarkTracing' BENCHTIMEOUT=50m > benchmark-tracing.txt
    $ cat benchmark-tracing.txt | grep -v tracing=t | sed 's/tracing=f//' > old.txt
    $ cat benchmark-tracing.txt | grep -v tracing=f | sed 's/tracing=t//' > new.txt
    $ benchstat old.txt new.txt

    name                                   old time/op    new time/op    delta
    Tracing/Cockroach//Scan1-24               382µs ± 2%     412µs ± 8%   +7.68%  (p=0.000 n=10+10)
    Tracing/Cockroach//Insert-24              579µs ± 2%     609µs ± 6%   +5.10%  (p=0.000 n=10+10)
    Tracing/MultinodeCockroach//Scan1-24      799µs ± 2%     885µs ± 1%  +10.76%  (p=0.000 n=10+9)
    Tracing/MultinodeCockroach//Insert-24    1.09ms ± 1%    1.14ms ± 2%   +5.04%  (p=0.000 n=9+10)

    name                                   old alloc/op   new alloc/op   delta
    Tracing/Cockroach//Scan1-24              25.3kB ± 5%    32.0kB ± 4%  +26.55%  (p=0.000 n=10+10)
    Tracing/Cockroach//Insert-24             38.0kB ± 6%    42.2kB ± 5%  +11.02%  (p=0.000 n=10+10)
    Tracing/MultinodeCockroach//Scan1-24     77.7kB ± 8%    97.1kB ±12%  +25.05%  (p=0.000 n=10+10)
    Tracing/MultinodeCockroach//Insert-24     107kB ± 8%     115kB ± 7%   +7.66%  (p=0.023 n=10+10)

    name                                   old allocs/op  new allocs/op  delta
    Tracing/Cockroach//Scan1-24                 256 ± 1%       280 ± 1%   +9.10%  (p=0.000 n=9+9)
    Tracing/Cockroach//Insert-24                301 ± 2%       321 ± 1%   +6.64%  (p=0.000 n=10+10)
    Tracing/MultinodeCockroach//Scan1-24        787 ± 2%       921 ± 2%  +17.04%  (p=0.000 n=9+9)
    Tracing/MultinodeCockroach//Insert-24       748 ± 1%       805 ± 2%   +7.61%  (p=0.000 n=9+9)

Release note: None

irfansharif committed:
[wip,dnm] tracing: retrieve log tags from parent span...

Instead of the incoming ctx. Going through the ctx package is expensive,
and log tags are unlikely to change between parent and child spans.

---

Actually, drop log tags altogether.

Release note: None

irfansharif committed:
[wip,dnm] tracing: drop span tags altogether

Release note: None

irfansharif committed:
[wip,dnm] tracing: more peephole optimizations

- Reduce lock contention by only deleting root spans from registry.
- Don't materialize span recordings unless absolutely needed.
- Compute span duration outside of mutex

Release note: None

irfansharif committed:
[wip,dnm] tracing: misc. clean ups

Tracer.AlwaysTrace is redundant. We can also avoid doing opentracing
stuff if we have no external sinks configured.

Release note: None

irfansharif committed:
[wip,dnm] execinfra: drop empty trace metadata

Release note: None

irfansharif committed:
[wip,dnm] tracing: pool SpanMeta objects

irfansharif committed:
[wip,dnm] rpc,exec: performant nodeid.string, skip metadata

irfansharif committed:
[wip,dnm] tracing: note down ideas around collapsing recordings

irfansharif committed:
[wip,dnm] tracing: collapse recordings if nothing is present

irfansharif committed:
[wip,dnm] tracing: collapse everything into the root span

irfansharif committed:
[wip,dnm] tracing: group allocation of span objects

irfansharif committed:
[wip,dnm] tracing: move from interface type; avoid string.ToLower

irfansharif committed:
Revert "[wip,dnm] tracing: group allocation of span objects"

This reverts commit 1631298012ccbe1bfb064698a36f510e519ce834.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62227)
<!-- Reviewable:end -->

irfansharif committed:
[wip,dnm] tracing: pool more meta objects,

- use an optimizedCtx for collapsed spans
- avoid stepping through interface for carrier
- mutate incoming ctx's metadata.MD directly
- disable emptry registry check, we broke it a few commits ago
([wip,dnm] tracing: move from interface type; avoid string.ToLower)
- skip some common span tags

Release note: None

irfansharif committed:
[wip,dnm] tracing: pool metadata.MD objects

irfansharif commented:
The sum of these changes look promising-ish. Here are 5 runs with kv95.
```
roachtest run kv95/enc=false/nodes=1/cpu=32  --workload ./workload-linux --cockroach ./cockroach-linux-62227 --count 5 --cpu-quota 320
roachtest run kv95/enc=false/nodes=1/cpu=32  --workload ./workload-linux --cockroach ./cockroach-linux-pre-62227 --count 5 --cpu-quota 320
```
```
# 62227 with sample_rate=0.1 vs. pre-62227 with sample_rate=0.1

# 62227
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       39558963        65931.4      0.8      0.7      1.6      3.3     31.5  read
  600.0s        0       38167539        63612.4      0.9      0.7      1.6      3.4     27.3  read
  600.0s        0       38060299        63433.8      0.9      0.8      1.6      3.5     32.5  read
  600.0s        0       39589903        65983.1      0.8      0.7      1.6      3.3     29.4  read
  600.0s        0       39046176        65076.9      0.9      0.8      1.6      3.4     30.4  read
  600.0s        0        2008289         3347.1      2.7      1.9      7.6     11.5     32.5  write
  600.0s        0        2083240         3472.1      2.3      1.8      5.5      8.9     30.4  write
  600.0s        0        2082753         3471.3      2.3      1.9      5.8      9.4     28.3  write
  600.0s        0        2053367         3422.3      2.2      1.7      5.5      8.9     28.3  write
  600.0s        0        2003421         3339.0      2.5      2.0      6.0      9.4     31.5  write

# pre-62227
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       36898941        61498.0      0.9      0.8      1.6      3.8     29.4  read
  600.0s        0       34131762        56886.2      1.0      0.9      1.8      4.5     39.8  read
  600.0s        0       37442791        62404.6      0.9      0.7      1.6      3.7     26.2  read
  600.0s        0       37646562        62744.1      0.9      0.8      1.6      3.7     35.7  read
  600.0s        0       38124218        63540.3      0.9      0.8      1.6      3.5     37.7  read
  600.0s        0        2004687         3341.1      2.4      1.9      6.0      9.4     29.4  write
  600.0s        0        1796091         2993.5      2.7      2.2      6.8     11.0     58.7  write
  600.0s        0        1940652         3234.4      2.5      2.0      6.3     10.0     28.3  write
  600.0s        0        1972014         3286.7      2.8      1.9      7.9     11.5     35.7  write
  600.0s        0        1980551         3300.9      2.5      2.0      6.3     10.0     35.7  write
median (read): 65076.90 vs. 62404.60 ops/sec ------ +4.28%
median (writ):  3422.30 vs.  3286.70 ops/sec ------ +4.13%
```

I'll work on polish in separate PRs.
